File: sorting/merge.cpp
#include <iostream>
using namespace std;
void merge(int *arr , int s , int e ){
     int mid = s + (e-s)/2;
     int len1 = mid - s +1;
     int len2 = e - mid;
     
    
    int *first = new int[len1];
    int *second = new int[len2];
    int mainindex=s;
    //copying values of 1st part
    for (int  i = 0; i < len1; i++)
    {
        first[i]=arr[mainindex++];
    }
    //copying values of 2nd part
    mainindex=mid+1;
    for (int  i = 0; i < len2; i++)
    {
        second[i]=arr[mainindex++];
    }
    
    //merge karde
    int index1=0;
    int index2=0;
    mainindex=s;
    while (index1<len1 && index2 <len2)
    {
        if (first[index1]<second[index2])
        {
            arr[mainindex++]=first[index1++];
        }
        else {
            arr[mainindex++]=second[index2++];
        }
        
    }
    
    while (index1<len1)
    {
        arr[mainindex++]=first[index1++];
    }
    while(index2<len2)
    {
        arr[mainindex++]=second[index2++];
    }
    
    delete []first;
    delete []second;
}
void mergesort(int *arr , int s , int e){
    if(s>=e){
        return;
    }
    int mid = s + (e-s)/2;
    //left
    mergesort(arr , s , mid );
    //right 
    mergesort(arr , mid +1 , e);

    merge (arr, s, e );

    
}
int main(){
    int arr[5]={5,4,3,2,1};
    int n = 5;
    mergesort(arr, 0 , n-1);

    for (int  i = 0; i < n ; i++)
    {
        cout<<arr[i]<<endl;
    }
    
}

File: sorting/merge.py
def merge(arr, s, e):
    mid = s + (e - s) // 2
    len1 = mid - s + 1
    len2 = e - mid
 
    first = arr[s : mid + 1]
    second = arr[mid + 1 : e + 1]
 
    i = j = 0
    k = s
    
    while i < len1 and j < len2:
        if first[i] < second[j]:
            arr[k] = first[i]
            i += 1
        else:
            arr[k] = second[j]
            j += 1
        k += 1
 
    while i < len1:
        arr[k] = first[i]
        i += 1
        k += 1
 
    while j < len2:
        arr[k] = second[j]
        j += 1
        k += 1
 
def merge_sort(arr, s, e):
    if s < e:
        mid = s + (e - s) // 2
        merge_sort(arr, s, mid)
        merge_sort(arr, mid + 1, e)
        merge(arr, s, e)
 

arr = [5, 4, 3, 2, 1]
n = len(arr)
merge_sort(arr, 0, n - 1)
print(arr)
